### DISCUSSION TITLE
Why 0.0 is not in the valid range of a float64 type (Bug?)

### DISCUSSION THREAD

**Discussion by IkerDC** [edgexfoundry/community#156] [Q&A]
## Context
I am using the `device-rest-go` to read some resources. One of them is a type Float64. I have stumbled upon some weird behavior, which seems like a bug to me, although I'll like to ask first before opening a PR (in case it is the expected behavior).

On the `driver.go` file, on the `HandleReadCommand()` function, each resources is validated.
```go
// driver.go
func (driver *RestDriver) HandleReadCommands(deviceName string, protocols map[string]models.ProtocolProperties, reqs []dsModels.CommandRequest) (responses []*dsModels.CommandValue, err error) {
	// ...
		if deviceResource.Properties.ValueType == common.ValueTypeBinary ||
			deviceResource.Properties.ValueType == common.ValueTypeObject {
			reading = body
		} else {
			reading = string(body)
		}
		contentType := resp.Header.Get(common.ContentType)

		val, err = validateCommandValue(deviceResource, reading, deviceResource.Properties.ValueType, contentType)
		if err != nil {
			return nil, fmt.Errorf("Recevice response data is not valid")
		}
        // ...
}
```
**Consider reading is a float64 equal to 0.0 `reading := 0.0`.**

Inside the validator, function (considering the reading ValueType is Float64), it will enter this case.
```go
// resthandler.go
func validateCommandValue(resource model.DeviceResource, reading interface{}, valueType string, contentType string) (interface{}, error) {
	switch valueType {
	case common.ValueTypeBinary:
		// ...
        case common.ValueTypeFloat64:
		val, err = cast.ToFloat64E(reading)
		if err != nil {
			return nil, fmt.Errorf(castError, resource.Name, err)
		}
		if err := checkFloatValueRange(valueType, val); err != nil {
			return nil, err
		}
        default:
                // ...
```  
However, the range validator function looks like this.
```go
// resthandler.go

func checkFloatValueRange(valueType string, val interface{}) error {
	switch valueType {
	case common.ValueTypeFloat32:
		valFloat := val.(float32)
		if math.Abs(float64(valFloat)) >= math.SmallestNonzeroFloat32 && math.Abs(float64(valFloat)) <= math.MaxFloat32 {
			return nil
		}
	case common.ValueTypeFloat64:
		valFloat := val.(float64)
		if math.Abs(valFloat) >= math.SmallestNonzeroFloat64 && math.Abs(valFloat) <= math.MaxFloat64 {
			return nil
		}
	}

	return fmt.Errorf("value %v for %s type is out of range", val, valueType)
}
```
To me, here there is a check missing for the case `valFloat == 0.0`. Note that if it is equal to 0 it will return an error: `"value 0 for Float64 type is out of range"`, because it is outside the range of `math.SmallestNonzeroFloat64 = 4.94066e-324` and `math.MaxFloat64 = 1.79769e+308`, but 0.0 is a valid float...

## Question.
1. Why the zero is not being considered?
2. Why is this range check being done? valFloat is being casted to a float64 in the line above the value shouldn't be already a valid float64 and in range? (otherwise the casting would fail). This is what is done for the `checkUintValueRange()/checkIntValueRange()`...

**Comment by cloudxxx8** [comment]
Yeah, it's a bug.
Thanks for reporting it.

**Reply by cloudxxx8** [reply]
@IkerDC please help open a bug in [device-rest-go](https://github.com/edgexfoundry/device-rest-go)
